### PR TITLE
Fix tokenizers caching

### PR DESCRIPTION
**Pull request by lhoestq**
I've found some cases where the caching didn't work properly for tokenizers:

1. if a tokenizer has a regex pattern, then the caching would be inconsistent across sessions
2. if a tokenizer has a cache attribute that changes after some calls, the the caching would not work after cache updates
3. if a tokenizer is used inside a function, the caching of this function would result in the same cache file for different tokenizers
4. if `unique_no_split_tokens`'s attribute is not the same across sessions (after loading a tokenizer) then the caching could be inconsistent

To fix that, this is what I did:

1. register a specific `save_regex` function for pickle that makes regex dumps deterministic
2. ignore cache attribute of some tokenizers before dumping
3. enable recursive dump by default for all dumps
4. make `unique_no_split_tokens` deterministic in https://github.com/huggingface/transformers/pull/6461

I also added tests to make sure that tokenizers hashing works as expected.
In the future we should find a way to test if hashing also works across session (maybe using two CI jobs ? or by hardcoding a tokenizer's hash ?)

### CODE REVIEW

**Comment by lhoestq** [comment]
This should fix #501 and also the issue you sent me on slack @sgugger .

File: src/nlp/utils/py_utils.py [hunk 1]
```diff
@@ -268,14 +270,29 @@ class Pickler(dill.Pickler):
...
+    Pickler(file, recurse=True).dump(obj)
     return
 
 
+@contextlib.contextmanager
+def _no_cache_fields(obj):
+    if _transformers_available:
+        import transformers as tr
+
+        if isinstance(obj, (tr.CTRLTokenizer, tr.GPT2Tokenizer, tr.OpenAIGPTTokenizer, tr.XLMTokenizer)):
```
Comment (thomwolf) [inline src/nlp/utils/py_utils.py:282]: This list might be a bit cumbersome to maintain in the future.
Should we just do a check that the `cache` attribute exists and is a dict maybe?

File: src/nlp/utils/py_utils.py [hunk 2]
```diff
@@ -297,9 +314,11 @@ def save_code(pickler, obj):
     dill._dill.log.info("Co: %s" % obj)
     # Filenames of functions created in notebooks or shells start with '<'
     # ex: <ipython-input-13-9ed2afe61d25> for ipython, and <stdin> for shell
+    # Moreover lambda functions have a special name: '<lambda>'
+    # ex: (lambda x: x).__code__.co_name == "<lambda>"  # True
     # Only those two lines are different from the original implementation:
-    co_filename = "" if obj.co_filename.startswith("<") else obj.co_filename
-    co_firstlineno = 1 if obj.co_filename.startswith("<") else obj.co_firstlineno
+    co_filename = "" if obj.co_filename.startswith("<") or obj.co_name == "<lambda>" else obj.co_filename
+    co_firstlineno = 1 if obj.co_filename.startswith("<") or obj.co_name == "<lambda>" else obj.co_firstlineno
```
Comment (thomwolf) [inline src/nlp/utils/py_utils.py:321]: Good!

File: src/nlp/utils/py_utils.py (hunk 1 above)
Comment (lhoestq) [reply src/nlp/utils/py_utils.py]: I changed it so that it checks if `cache` exists and is a dict

**Review by thomwolf — APPROVED** [review]
Good to merge for me when you are happy with it :-)

**Pull request merged**